SYCL/ESIMD/api/functional/ctors/common.hpp

```diff
@@ -0,0 +1,165 @@
+
+namespace esimd_test {
+namespace api {
+namespace functional {
+
+// Dummy kernel for submitting some code into device side.
+template <typename DataT, int NumElems, typename T> struct Kernel;
+
+template <typename DataT>
+using allocator_type = sycl::usm_allocator<DataT, sycl::usm::alloc::shared>;
+
+template <typename DataT>
+using vector_with_allocator = std::vector<DataT, allocator_type<DataT>>;
+
+// Struct that provise methods for constructing inf, lowest, nan and max values
+// for reference data.
+template <typename DataT> struct value {
+  static DataT inf() {
+    DataT inf_val{};
+    if constexpr (std::is_same_v<DataT, sycl::half>) {
+      inf_val = 0;
+    } else {
+      inf_val = std::numeric_limits<DataT>::infinity();
+    }
+    return inf_val;
+  }
+
+  static DataT lowest() {
+    DataT min_val{};
+    if constexpr (std::is_same_v<DataT, sycl::half>) {
+      min_val = 0;
+    } else {
+      min_val = std::numeric_limits<DataT>::lowest();
+    }
+    return min_val;
+  }
+
+  static DataT nan() {
+    DataT nan_val{};
+    if constexpr (std::is_same_v<DataT, double>) {
+      nan_val = sycl::nan(42UL);
+    } else if constexpr (std::is_same_v<DataT, float>) {
+      nan_val = sycl::nan(42U);
+    } else if constexpr (std::is_same_v<DataT, sycl::half>) {
+      nan_val = 0;
+    }
+    return nan_val;
+  }
+
+  static DataT max() {
+    DataT max_val{};
+    if constexpr (std::is_same_v<DataT, sycl::half>) {
+      max_val = 0;
+    } else {
+      max_val = std::numeric_limits<DataT>::max();
+    }
+    return max_val;
+  }
+};
+
+// Generate std::vector with reference according to current tested data type and
+// number of elements.
+template <typename DataT, int NumElems>
+void generate_ref_data(vector_with_allocator<DataT> &ref_data) {
+
+  const DataT min = value<DataT>::lowest();
+  const DataT min_half = min / 2;
+  const DataT max = value<DataT>::max();
+  const DataT max_half = max / 2;
+  const DataT inf = value<DataT>::inf();
+  const DataT min_plus_one = min + 1;
+  const DataT max_minus_one = max - 1;
+
+  if constexpr (std::is_signed_v<DataT>) {
+    if constexpr (NumElems == 1) {
+
+      ref_data.insert(ref_data.end(), {min, min_half, max, max_half, 0});
+    } else {
+      ref_data.insert(ref_data.end(), {min, min_plus_one, min_half, max,
+                                       max_minus_one, max_half, -1, 0});
+      for (size_t it = 8; it < NumElems; it++) {
+        ref_data.push_back(it);
+      }
+    }
+  }
+
+  else if constexpr (std::is_unsigned_v<DataT>) {
+    if constexpr (NumElems == 1) {
+      ref_data.insert(ref_data.end(), {max, max_half, 0});
+    } else {
+      ref_data.insert(ref_data.end(), {max, max_minus_one, max_half, 0});
+      for (size_t it = 4; it < NumElems; it++) {
+        ref_data.push_back(it);
+      }
+    }
+  }
+
+  else if constexpr (std::is_same_v<DataT, sycl::half> ||
+                     std::is_floating_point_v<DataT>) {
+    DataT nan{value<DataT>::nan()};
```

Comment:
```suggestion
    static const DataT nan{value<DataT>::nan()};
```
The idea behind the changes is as follows:
(1) const-correctness guarantees we don't change this value later, so it's more like an alias than like a variable
(2) static within the function scope enables the lazy-initialization - with no need to re-initialize with every call of `generate_ref_data`
I would suggest to apply such changes for all `value<DataT>::` synonyms in this function

Reply:
I've done `static_assert` in this commit: 74c17944c3316f44492338a96ffa9d2ec008a0c5